subprojects/build-events/src/main/java/org/gradle/internal/build/event/types/DefaultFailure.java

```diff
@@ -81,26 +82,29 @@ public static InternalFailure fromThrowable(Throwable t, ProblemLookup problemLo
         StringWriter out = new StringWriter();
         PrintWriter wrt = new PrintWriter(out);
         t.printStackTrace(wrt);
-        Throwable cause = t.getCause();
-        List<InternalFailure> causeFailures;
+        List<InternalFailure> causeFailures = getCauseFailures(problemLookup, mapper, t.getCause());
+
+        List<InternalBasicProblemDetailsVersion3> problemDetails = problemLookup.findAll(t).stream()
+            .map(mapper)
+            .filter(Objects::nonNull)
+            .collect(toImmutableList());
+        return new DefaultFailure(t.getMessage(),
+            out.toString(),
+            causeFailures,
+            problemDetails);
+    }
+
+    @Nonnull
+    private static List<InternalFailure> getCauseFailures(ProblemLookup problemLookup, Function<Problem, InternalBasicProblemDetailsVersion3> mapper, Throwable cause) {
         if (cause == null) {
-            causeFailures = Collections.emptyList();
+            return ImmutableList.of();
         } else if (cause instanceof MultiCauseException) {
             MultiCauseException multiCause = (MultiCauseException) cause;
-            causeFailures = multiCause.getCauses().stream().map(f -> fromThrowable(f, problemLookup, mapper)).collect(Collectors.toList());
-        } else {
-            causeFailures = Collections.singletonList(fromThrowable(cause, problemLookup, mapper));
-        }
-        Collection<Problem> problemMapping = problemLookup.findAll(t);
-
-        List<Problem> problems = new ArrayList<>();
-        if (problemMapping != null) {
-            problems.addAll(problemMapping);
-        }
-        if (problems.isEmpty()) {
```

Comment:
It appears to me, this doesn't make a difference. it will get assigned an empty list in the constructor anyway.
If it does we need to comment why.